File: api/src/main/java/net/blueshell/api/validation/user/ExistingUsernameValidator.java
package net.blueshell.api.validation.user;

import jakarta.validation.ConstraintValidator;
import jakarta.validation.ConstraintValidatorContext;
import net.blueshell.api.service.UserService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;
import org.springframework.util.StringUtils;

/**
 * Validator to check if the username is unique.
 */
@Component
public class ExistingUsernameValidator implements ConstraintValidator<UniqueUsername, String> {

    private final UserService userService;

    @Autowired
    public ExistingUsernameValidator(UserService userService) {
        this.userService = userService;
    }

    @Override
    public boolean isValid(String username, ConstraintValidatorContext context) {
        if (!StringUtils.hasText(username)) {
            // Let @NotBlank handle this
            return true;
        }
        return userService.existsByUsername(username);
    }
}

File: api/src/main/kotlin/net/blueshell/api/validation/user/ExistingUsernameValidator.kt
package net.blueshell.api.validation.user

import jakarta.validation.ConstraintValidator
import jakarta.validation.ConstraintValidatorContext
import net.blueshell.api.service.UserService
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component
import org.springframework.util.StringUtils

/**
 * Validator to check if the username is unique.
 */
@Component
class ExistingUsernameValidator @Autowired constructor(private val userService: UserService) :
    ConstraintValidator<ExistingUsername?, String?> {
    override fun isValid(username: String?, context: ConstraintValidatorContext?): Boolean {
        if (!StringUtils.hasText(username)) {
            // Let @NotBlank handle this
            return true
        }
        return userService.existsByUsername(username)
    }
}
